plugins/org.komodo.relational.commands/src/org/komodo/relational/commands/RelationalLabelProvider.java

```diff
@@ -50,182 +48,35 @@ public RelationalLabelProvider() {
-                        }
-                        continue;
-                    }
-                }
-
-                if ( this.status.isShowingPropertyNamePrefixes() ) {
-                    if ( parent.hasChild( this.status.getTransaction(), segment ) ) {
-                        kobject = parent.getChild( this.status.getTransaction(), segment );
-                        parent = kobject;
-                    } else {
-                        return null; // no child with that name
-                    }
-                } else {
-                    // loop through children and take first one with local name match.  skip grouping nodes
-                    boolean childFound = false;
-                    for ( final KomodoObject kid : parent.getChildren( this.status.getTransaction() ) ) {
-                        final String name = kid.getName( this.status.getTransaction() );
-                        final int index = name.indexOf( StringConstants.COLON );
-
-                        if ( index == -1 ) {
-                            if ( segment.equals( name ) ) {
-                                kobject = kid;
-                                parent = kobject;
-                                childFound = true;
-                                break;
-                            }
-                        } else if (GROUPING_NODES.contains( name )) {
-                            KomodoObject[] children = kid.getChildren(this.status.getTransaction(), segment);
-                            if(children.length>0) {
-                                kobject = children[0];
-                                parent = kobject;
-                                childFound = true;
-                                break;
-                            }
-                        } else {
-                            if ( segment.equals( name.substring( index + 1 ) ) ) {
-                                kobject = kid;
-                                parent = kobject;
-                                childFound = true;
-                                break;
-                            }
-                        }
-                    }
-                    if(!childFound) return null;
-                }
-            }
-
-            if ( kobject == null ) {
-                return null;
-            }
-
-            return kobject.getAbsolutePath();
-        } catch ( final Exception e ) {
-            return null;
-        }
-    }
+	@Override
+	public List<String> getGroupingNodes(){
+		return GROUPING_NODES;
+	}
+	
+	/**
+	 * {@inheritDoc}
+	 * @throws KException 
+	 * 
+	 * @see org.komodo.shell.api.KomodoObjectLabelProvider#getTypeDisplay(org.komodo.spi.repository.Repository.UnitOfWork,
+	 *      org.komodo.spi.repository.KomodoObject)
+	 */
+	@Override
+	public String getTypeDisplay(UnitOfWork uow, KomodoObject kobject){
+		for(ShellCommandProvider provider:status.getCommandFactory().getCommandProviders()){
```

Comment:
You are looping through *all* providers, even those that are not all relational. And each provider has to `resolve` the object. Maybe we should find a more efficient way. One way might be to have an interface that each of the relational  `ShellCommandProviders` implement. Then you could only ask those who inherit that interface to resolve the object. Then you would return once you resolved to a non-null object. This interface could have `default` implementations of the `getStatusMessage` method (return `null`) and the `initWorkspaceState` method (no method body). There might be other ways to short circuit this. Wdyt?